# Prediction model: Logistic Regression

In [4]:
import numpy as np
import pandas as pd
from pandas import DataFrame, Series

In [5]:
kobe_num = pd.DataFrame.from_csv("./data/kobe_num_variables.csv", sep=",")

In [6]:
kobe_num.shape

(30697, 17)

In [7]:
kobe_num.head(3)

game_event_id   game_id  loc_x  loc_y  period  playoffs  \
shot_id                                                            
1                   10  20000012    167     72       1         0   
2                   12  20000012   -157      0       1         0   
3                   35  20000012   -101    135       1         0   

         shot_distance  shot_made_flag  seconds_to_period_end  \
shot_id                                                         
1                   18             NaN                    627   
2                   15               0                    622   
3                   16               1                    465   

         accurate_shot_distance  game_year  game_month  game_dayofweek  \
shot_id                                                                  
1                          18.2       2000          10               1   
2                          15.7       2000          10               1   
3                          16.9       2000          10               1   

         game_dayofyear  game_dayofweek.1  game_dayofyear.1  local  
shot_id                                                             
1                   305                 1               305      0  
2                   305                 1               305      0  
3                   305                 1               305      0

In [8]:
kobe_old_cat = pd.DataFrame.from_csv("./data/kobe_old_cat_variables.csv", sep=",")

In [9]:
kobe_old_cat.shape

(30697, 114)

In [10]:
kobe_old_cat.head(3)

action_type_Alley Oop Dunk Shot  action_type_Alley Oop Layup shot  \
shot_id                                                                      
1                                      0                                 0   
2                                      0                                 0   
3                                      0                                 0   

         action_type_Cutting Finger Roll Layup Shot  \
shot_id                                               
1                                                 0   
2                                                 0   
3                                                 0   

         action_type_Cutting Layup Shot  action_type_Driving Bank shot  \
shot_id                                                                  
1                                     0                              0   
2                                     0                              0   
3                                     0                              0   

         action_type_Driving Dunk Shot  \
shot_id                                  
1                                    0   
2                                    0   
3                                    0   

         action_type_Driving Finger Roll Layup Shot  \
shot_id                                               
1                                                 0   
2                                                 0   
3                                                 0   

         action_type_Driving Finger Roll Shot  \
shot_id                                         
1                                           0   
2                                           0   
3                                           0   

         action_type_Driving Floating Bank Jump Shot  \
shot_id                                                
1                                                  0   
2                                                  0   
3                                                  0   

         action_type_Driving Floating Jump Shot      ...       opponent_PHI  \
shot_id                                              ...                      
1                                             0      ...                  0   
2                                             0      ...                  0   
3                                             0      ...                  0   

         opponent_PHX  opponent_POR  opponent_SAC  opponent_SAS  opponent_SEA  \
shot_id                                                                         
1                   0             1             0             0             0   
2                   0             1             0             0             0   
3                   0             1             0             0             0   

         opponent_TOR  opponent_UTA  opponent_VAN  opponent_WAS  
shot_id                                                          
1                   0             0             0             0  
2                   0             0             0             0  
3                   0             0             0             0  

[3 rows x 114 columns]

In [11]:
kobe_new_cat = pd.DataFrame.from_csv("./data/kobe_new_cat_variables.csv", sep=",")

In [12]:
kobe_new_cat.shape

(30697, 871)

In [13]:
kobe_new_cat.head(3)

at_sza_Alley Oop Dunk Shot-Center(C)  \
shot_id                                         
1                                           0   
2                                           0   
3                                           0   

         at_sza_Alley Oop Layup shot-Center(C)  \
shot_id                                          
1                                            0   
2                                            0   
3                                            0   

         at_sza_Cutting Finger Roll Layup Shot-Center(C)  \
shot_id                                                    
1                                                      0   
2                                                      0   
3                                                      0   

         at_sza_Cutting Layup Shot-Center(C)  \
shot_id                                        
1                                          0   
2                                          0   
3                                          0   

         at_sza_Driving Bank shot-Center(C)  \
shot_id                                       
1                                         0   
2                                         0   
3                                         0   

         at_sza_Driving Dunk Shot-Center(C)  \
shot_id                                       
1                                         0   
2                                         0   
3                                         0   

         at_sza_Driving Finger Roll Layup Shot-Center(C)  \
shot_id                                                    
1                                                      0   
2                                                      0   
3                                                      0   

         at_sza_Driving Finger Roll Shot-Center(C)  \
shot_id                                              
1                                                0   
2                                                0   
3                                                0   

         at_sza_Driving Floating Bank Jump Shot-Left Side(L)  \
shot_id                                                        
1                                                        0     
2                                                        0     
3                                                        0     

         at_sza_Driving Floating Jump Shot-Center(C)  \
shot_id                                                
1                                                  0   
2                                                  0   
3                                                  0   

                     ...               at_sd_Turnaround Jump Shot-25  \
shot_id              ...                                               
1                    ...                                           0   
2                    ...                                           0   
3                    ...                                           0   

         at_sd_Turnaround Jump Shot-26  at_sd_Turnaround Jump Shot-27  \
shot_id                                                                 
1                                    0                              0   
2                                    0                              0   
3                                    0                              0   

         at_sd_Turnaround Jump Shot-3  at_sd_Turnaround Jump Shot-4  \
shot_id                                                               
1                                   0                             0   
2                                   0                             0   
3                                   0                             0   

         at_sd_Turnaround Jump Shot-5  at_sd_Turnaround Jump Shot-6  \
shot_id                                                               
1                                   0                             0   
2                           

## Splitting the submission data

In [14]:
kobe_num_submission = kobe_num[pd.isnull(kobe_num["shot_made_flag"])]

We add 'shot_made_flag' to 'kobe_old_cat' and 'kobe_new_cat'

In [15]:
kobe_target = kobe_num["shot_made_flag"]

In [16]:
kobe_old_cat = pd.concat([kobe_old_cat, kobe_target], axis=1)

In [17]:
kobe_new_cat = pd.concat([kobe_new_cat, kobe_target], axis=1)

And we create the two remaining submission data

In [18]:
kobe_old_cat_submission = kobe_old_cat[pd.isnull(kobe_old_cat["shot_made_flag"])]

In [182]:
kobe_new_cat_submission = kobe_new_cat[pd.isnull(kobe_new_cat["shot_made_flag"])]

In [183]:
print kobe_num_submission.shape
print kobe_old_cat_submission.shape
print kobe_new_cat_submission.shape

(5000, 17)
(5000, 115)
(5000, 872)


We split the dataset from submission data

In [184]:
kobe_num = kobe_num[pd.notnull(kobe_num["shot_made_flag"])]
kobe_old_cat = kobe_old_cat[pd.notnull(kobe_old_cat["shot_made_flag"])]
kobe_new_cat = kobe_new_cat[pd.notnull(kobe_new_cat["shot_made_flag"])]

In [185]:
print len(kobe_num)
print len(kobe_old_cat)
print len(kobe_new_cat)

25697
25697
25697


## Logistic Regression

Finally we finally run logistic regression on the three data set ('kobe_num', 'kobe_old_cat', 'kobe_new_cat'). Let see if we can find an aggregation that can be accurate.

In [186]:
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn import metrics
from sklearn.cross_validation import cross_val_score

#### Let's start with the first data, that should be not particuarly accurate.

We need to flatten `y` into a 1-D array, so that scikit-learn will properly understand it as the response variable.

In [187]:
y = np.ravel(kobe_num["shot_made_flag"])

In [188]:
kobe_num = kobe_num.drop("shot_made_flag", axis=1)

In [189]:
X = kobe_num

In [190]:
model = LogisticRegression()
model = model.fit(X, y)

In [191]:
model.score(X, y)

0.55383896952951706

55.4% accuracy doesn't particularly seem good.

In [192]:
y.mean()

0.44616103047048294

44.6% of the shots were in. We could obtain 68% accuracy by always predicting "no".

Let examine the coefficients to see what we learn.

In [193]:
pd.DataFrame(zip(X.columns, np.transpose(model.coef_)))

0                     1
0            game_event_id   [-1.2047203577e-13]
1                  game_id   [-8.0092441746e-09]
2                    loc_x   [-3.1849889696e-15]
3                    loc_y  [-1.09848332124e-13]
4                   period  [-1.10112622468e-15]
5                 playoffs  [-2.40143249124e-17]
6            shot_distance  [-1.58293308464e-14]
7    seconds_to_period_end  [-7.39248545295e-14]
8   accurate_shot_distance  [-1.61601027326e-14]
9                game_year  [-7.01575439594e-13]
10              game_month  [-2.08731855382e-15]
11          game_dayofweek  [-1.14664000538e-15]
12          game_dayofyear  [-5.79632862635e-14]
13        game_dayofweek.1  [-1.14664000538e-15]
14        game_dayofyear.1  [-5.79632862635e-14]
15                   local  [-1.10021473863e-16]

## Model Evaluation using a validation set

So far we have trained and tested on the same set. Now we split the data into a training set and a testing set to evaluate the model.

In [194]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
model2 = LogisticRegression()
model2.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

Now we predict the class labels for the test set.

In [195]:
predicted = model2.predict(X_test)
print predicted

[ 0.  0.  0. ...,  0.  0.  0.]


We will also generate the class probabilities.

In [196]:
probs = model2.predict_proba(X_test)
print probs

[[ 0.59550493  0.40449507]
 [ 0.54087755  0.45912245]
 [ 0.57761485  0.42238515]
 ..., 
 [ 0.54145566  0.45854434]
 [ 0.53952791  0.46047209]
 [ 0.5406845   0.4593155 ]]


The classifier is predicting a 1 (shot made) any time the probability in the second column is greater than 0.5.

Now we generate some evaluation metrics.

In [197]:
print metrics.accuracy_score(y_test, predicted)
print metrics.roc_auc_score(y_test, probs[:, 1])

0.558236057069
0.505855857638


The accuracy is 55.8%, which is similar to what we experienced when training and predicting on the same data.

We can also see the confusion matrix and a classification report with other metrics.

In [198]:
print metrics.confusion_matrix(y_test, predicted)
print metrics.classification_report(y_test, predicted)

[[4304    0]
 [3406    0]]
             precision    recall  f1-score   support

        0.0       0.56      1.00      0.72      4304
        1.0       0.00      0.00      0.00      3406

avg / total       0.31      0.56      0.40      7710



## Model Evaluation Using Cross-validation

Now we evaluate the model using 10-fold cross-validation, to see if the accuracy holds up more rigorously.

In [199]:
scores = cross_val_score(LogisticRegression(), X, y, scoring='accuracy', cv=10)
print scores
print scores.mean()

[ 0.55387009  0.55387009  0.5536965   0.5536965   0.5536965   0.55391203
  0.55391203  0.55391203  0.55391203  0.55391203]
0.553838981321


At least it's still performing at around 55.4 accuracy.

#### Let's try with the sparse matrix of the original and theoretically more relevant variables.

In [200]:
y = np.ravel(kobe_old_cat["shot_made_flag"])

In [201]:
kobe_old_cat = kobe_old_cat.drop("shot_made_flag", axis =1)

In [202]:
X = kobe_old_cat

In [203]:
model = LogisticRegression()
model = model.fit(X, y)
model.score(X, y)

0.6821418842666459

68.2% accuracy. It seems that these variables are more useful.

In [205]:
pd.DataFrame(zip(X.columns, np.transpose(model.coef_)))

0                   1
0                action_type_Alley Oop Dunk Shot    [0.364298275533]
1               action_type_Alley Oop Layup shot    [0.115555123239]
2     action_type_Cutting Finger Roll Layup Shot               [0.0]
3                 action_type_Cutting Layup Shot   [0.0176686694331]
4                  action_type_Driving Bank shot   [-0.176419586754]
5                  action_type_Driving Dunk Shot     [1.12114890087]
6     action_type_Driving Finger Roll Layup Shot     [1.07552854665]
7           action_type_Driving Finger Roll Shot    [0.885420380264]
8    action_type_Driving Floating Bank Jump Shot    [0.297348061831]
9         action_type_Driving Floating Jump Shot   [-0.505905299936]
10                 action_type_Driving Hook Shot   [0.0179257044883]
11                 action_type_Driving Jump shot   [-0.669533563248]
12                action_type_Driving Layup Shot    [0.259723501046]
13        action_type_Driving Reverse Layup Shot    [0.316684295303]
14            action_type_Driving Slam Dunk Shot    [0.764068645887]
15                         action_type_Dunk Shot    [-1.15442339069]
16                action_type_Fadeaway Bank shot     [1.14845752068]
17                action_type_Fadeaway Jump Shot   [-0.391119037284]
18            action_type_Finger Roll Layup Shot    [0.634073767352]
19                  action_type_Finger Roll Shot   [-0.782425605274]
20                action_type_Floating Jump shot    [0.392048564884]
21               action_type_Follow Up Dunk Shot  [-0.0928028223037]
22                    action_type_Hook Bank Shot     [0.70606119597]
23                         action_type_Hook Shot    [-1.02823484406]
24                    action_type_Jump Bank Shot    [0.647501844089]
25                    action_type_Jump Hook Shot    [0.402184801451]
26                         action_type_Jump Shot    [-1.47517641738]
27                        action_type_Layup Shot    [-1.26879649014]
28                  action_type_Pullup Bank shot   [-0.734834931968]
29                  action_type_Pullup Jump shot    [0.213659268675]
..                                           ...                 ...
84                                  opponent_CHA   [0.0633824212377]
85                                  opponent_CHI   [0.0337721142242]
86                                  opponent_CLE   [0.0216599015715]
87                                  opponent_DAL   [0.0594432701699]
88                                  opponent_DEN   [0.0579063688816]
89                                  opponent_DET    [0.059549967497]
90                                  opponent_GSW  [-0.0448920253988]
91                                  opponent_HOU  [-0.0837747780867]
92                                  opponent_IND  [-0.0611729443001]
93                                  opponent_LAC   [0.0369248443816]
94                                  opponent_MEM  [-0.0194941602651]
95                                  opponent_MIA  [-0.0334183182468]
96                                  opponent_MIL   [0.0566628866813]
97                                  opponent_MIN   [0.0301064600749]
98                                  opponent_NJN   [-0.139022402278]
99                                  opponent_NOH   [0.0882557804028]
100                                 opponent_NOP   [0.0122117660598]
101                                 opponent_NYK    [0.256828215644]
102                                 opponent_OKC   [-0.222020596535]
103                                 opponent_ORL  [-0.0383175360413]
104                                 opponent_PHI   [0.0577875681672]
105                                 opponent_PHX    [0.163131577784]
106                                 opponent_POR   [0.0590948858336]
107                                 opponent_SAC    [0.154034409788]
108                                 opponent_SAS    [0.039654009052]
109                                 opponent_SEA    [0.170553253748]
110                                 opponent_

In [206]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
model2 = LogisticRegression()
model2.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [207]:
predicted = model2.predict(X_test)
print predicted

[ 0.  0.  0. ...,  0.  0.  1.]


In [208]:
probs = model2.predict_proba(X_test)
print probs

[[ 0.71035482  0.28964518]
 [ 0.64705777  0.35294223]
 [ 0.76465764  0.23534236]
 ..., 
 [ 0.67969767  0.32030233]
 [ 0.70420902  0.29579098]
 [ 0.21980381  0.78019619]]


In [209]:
print metrics.accuracy_score(y_test, predicted)
print metrics.roc_auc_score(y_test, probs[:, 1])

0.67924773022
0.680867133661


Similar accuracy to what we experienced when training and predicting on the same data.

In [210]:
print metrics.confusion_matrix(y_test, predicted)
print metrics.classification_report(y_test, predicted)

[[3682  622]
 [1851 1555]]
             precision    recall  f1-score   support

        0.0       0.67      0.86      0.75      4304
        1.0       0.71      0.46      0.56      3406

avg / total       0.69      0.68      0.66      7710



At least this model it's able to predict shot in with some accuracy

In [211]:
scores = cross_val_score(LogisticRegression(), X, y, scoring='accuracy', cv=10)
print scores
print scores.mean()

[ 0.6907818   0.68728121  0.68793774  0.69143969  0.69066148  0.67730634
  0.65940054  0.66952121  0.68703776  0.67497081]
0.681633858897


Fine. It's still performing at around 68% accuracy.

#### Now we'll try with the sparse matrix of the variables newly created crossing types of actions with different segmentation of position on court. There are 871 variables: without a feature selection it's highly probable the model will be not accurate

In [212]:
y = np.ravel(kobe_new_cat["shot_made_flag"])

In [213]:
kobe_new_cat = kobe_new_cat.drop("shot_made_flag", axis=1)

In [214]:
X = kobe_new_cat

In [215]:
model = LogisticRegression()
model = model.fit(X, y)
model.score(X, y)

0.6851772580456863

In [217]:
pd.DataFrame(zip(X.columns, np.transpose(model.coef_)))

0                    1
0                 at_sza_Alley Oop Dunk Shot-Center(C)     [0.698272680426]
1                at_sza_Alley Oop Layup shot-Center(C)    [0.0633427127911]
2      at_sza_Cutting Finger Roll Layup Shot-Center(C)                [0.0]
3                  at_sza_Cutting Layup Shot-Center(C)    [0.0131659837808]
4                   at_sza_Driving Bank shot-Center(C)   [-0.0157328545906]
5                   at_sza_Driving Dunk Shot-Center(C)     [0.900360619276]
6      at_sza_Driving Finger Roll Layup Shot-Center(C)     [0.314017373662]
7            at_sza_Driving Finger Roll Shot-Center(C)     [0.287451851291]
8    at_sza_Driving Floating Bank Jump Shot-Left Si...     [0.189826782984]
9          at_sza_Driving Floating Jump Shot-Center(C)   [-0.0646000710999]
10     at_sza_Driving Floating Jump Shot-Right Side(R)    [-0.331825922509]
11                  at_sza_Driving Hook Shot-Center(C)    [-0.112938254965]
12                  at_sza_Driving Jump shot-Center(C)   [-0.0481143753994]
13               at_sza_Driving Jump shot-Left Side(L)  [-0.00189682114857]
14              at_sza_Driving Jump shot-Right Side(R)    [-0.642302527103]
15                 at_sza_Driving Layup Shot-Center(C)   [-0.0582640208341]
16         at_sza_Driving Reverse Layup Shot-Center(C)     [0.103252415569]
17             at_sza_Driving Slam Dunk Shot-Center(C)      [0.74211772982]
18                          at_sza_Dunk Shot-Center(C)     [0.141500176278]
19                       at_sza_Dunk Shot-Left Side(L)     [0.189826782984]
20                 at_sza_Fadeaway Bank shot-Center(C)    [0.0840712878729]
21              at_sza_Fadeaway Bank shot-Left Side(L)     [0.141928413368]
22             at_sza_Fadeaway Bank shot-Right Side(R)     [0.531405216085]
23                 at_sza_Fadeaway Jump Shot-Center(C)       [0.1320788323]
24      at_sza_Fadeaway Jump Shot-Left Side Center(LC)    [-0.191154354541]
25              at_sza_Fadeaway Jump Shot-Left Side(L)    [-0.450069796586]
26     at_sza_Fadeaway Jump Shot-Right Side Center(RC)     [0.169042155357]
27             at_sza_Fadeaway Jump Shot-Right Side(R)     [-0.32931682505]
28             at_sza_Finger Roll Layup Shot-Center(C)     [0.224902742277]
29                   at_sza_Finger Roll Shot-Center(C)    [-0.402561629448]
..                                                 ...                  ...
841                       at_sd_Turnaround Hook Shot-5     [0.246759413671]
842                       at_sd_Turnaround Hook Shot-6   [-0.0690055132904]
843                       at_sd_Turnaround Hook Shot-8    [-0.290119076719]
844                       at_sd_Turnaround Hook Shot-9   [-0.0331031268205]
845                      at_sd_Turnaround Jump Shot-10     [-0.35868447389]
846                      at_sd_Turnaround Jump Shot-11   [-0.0978199497543]
847                      at_sd_Turnaround Jump Shot-12    [-0.123043038055]
848                      at_sd_Turnaround Jump Shot-13   [-0.0799762171736]
849                      at_sd_Turnaround Jump Shot-14     [0.213416025278]
850                      at_sd_Turnaround Jump Shot-15    [-0.442778558951]
851                      at_sd_Turnaround Jump Shot-16    [-0.411028209604]
852                      at_sd_Turnaround Jump Shot-17    [-0.165804001739]
853                      at_sd_Turnaround Jump Shot-18    [-0.375815017186]
854                      at_sd_Turnaround Jump Shot-19   [-0.0977074624009]
855                       at_sd_Turnaround Jump Shot-2                [0.0]
856                      at_sd_Turnaround Jump Shot-20     [0.297778138878]
857                      at_sd_Turnaround Jump Shot-21     [0.948858058317]
858                      at_sd_Turnaround Jump Shot-22     [0.264381578832]
859                      at_sd_Turnaround Jump Shot-23    [-0.248686045869]
860                      at_sd_Turnaround Jump Shot-24     [0.824664561966]
861                      at_sd_Turnaround Jump Shot-25    [-0.643307476344]
862                      

In [218]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
model2 = LogisticRegression()
model2.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [219]:
predicted = model2.predict(X_test)
print predicted

[ 0.  0.  0. ...,  0.  0.  1.]


In [220]:
probs = model2.predict_proba(X_test)
print probs

[[ 0.65403969  0.34596031]
 [ 0.67550001  0.32449999]
 [ 0.73605493  0.26394507]
 ..., 
 [ 0.67498036  0.32501964]
 [ 0.67803397  0.32196603]
 [ 0.41834019  0.58165981]]


Something less accurate then the second model.

In [221]:
print metrics.accuracy_score(y_test, predicted)
print metrics.roc_auc_score(y_test, probs[:, 1])

0.674967574578
0.686875214197


In [222]:
print metrics.confusion_matrix(y_test, predicted)
print metrics.classification_report(y_test, predicted)

[[3723  581]
 [1925 1481]]
             precision    recall  f1-score   support

        0.0       0.66      0.87      0.75      4304
        1.0       0.72      0.43      0.54      3406

avg / total       0.69      0.67      0.66      7710



In [223]:
scores = cross_val_score(LogisticRegression(), X, y, scoring='accuracy', cv=10)
print scores
print scores.mean()

[ 0.68650331  0.67989109  0.68365759  0.68677043  0.68599222  0.66757493
  0.65784352  0.66718568  0.68703776  0.67068898]
0.677314550057


#### The second model was a little bit better than the third. Let's try to merge all the variables formerly present in the dataset to see if the model manages to be tighter.

In [247]:
kobe_num = pd.DataFrame.from_csv("./data/kobe_num_variables.csv", sep=",")

In [248]:
kobe_old_cat = pd.DataFrame.from_csv("./data/kobe_old_cat_variables.csv", sep=",")

In [249]:
kobe_num = kobe_num[pd.notnull(kobe_num["shot_made_flag"])]

In [250]:
kobe = pd.concat([kobe_num, kobe_old_cat], axis=1)

In [251]:
kobe_submission = kobe[pd.isnull(kobe["shot_made_flag"])]

In [252]:
kobe = kobe[pd.notnull(kobe["shot_made_flag"])]

In [253]:
y = np.ravel(kobe["shot_made_flag"])

In [254]:
kobe = kobe.drop("shot_made_flag", axis=1)

In [255]:
X = kobe

In [256]:
model = LogisticRegression()
model = model.fit(X, y)
model.score(X, y)

0.55383896952951706

In [259]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
model2 = LogisticRegression()
model2.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [260]:
predicted = model2.predict(X_test)
print predicted

[ 0.  0.  0. ...,  0.  0.  0.]


In [261]:
probs = model2.predict_proba(X_test)
print probs

[[ 0.59550493  0.40449507]
 [ 0.54087755  0.45912245]
 [ 0.57761485  0.42238515]
 ..., 
 [ 0.54145566  0.45854434]
 [ 0.53952791  0.46047209]
 [ 0.5406845   0.4593155 ]]


In [262]:
print metrics.accuracy_score(y_test, predicted)
print metrics.roc_auc_score(y_test, probs[:, 1])

0.558236057069
0.505855857638


In [263]:
print metrics.confusion_matrix(y_test, predicted)
print metrics.classification_report(y_test, predicted)

[[4304    0]
 [3406    0]]
             precision    recall  f1-score   support

        0.0       0.56      1.00      0.72      4304
        1.0       0.00      0.00      0.00      3406

avg / total       0.31      0.56      0.40      7710



In [264]:
scores = cross_val_score(LogisticRegression(), X, y, scoring='accuracy', cv=10)
print scores
print scores.mean()

[ 0.55387009  0.55387009  0.5536965   0.5536965   0.5536965   0.55391203
  0.55391203  0.55391203  0.55391203  0.55391203]
0.553838981321


Definitely not a good model. Now we save the submission dataset corresponding to the second dataset to recreate the model in the following script and predict the probability of made shot on submission dataset.

In [19]:
kobe_old_cat_submission.to_csv("./data/kobe_model_submission.csv", sep= ",")